src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.Object.cs

```diff
@@ -0,0 +1,66 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if !NETCOREAPP
+using System.Diagnostics;
+#endif
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    public sealed partial class JsonContent
+    {
+        private sealed class JsonContentObjectSerializer : JsonContentSerializer
+        {
+            private readonly JsonSerializerOptions? _jsonSerializerOptions;
+
+            public override Type ObjectType { get; }
+            public override object? Value { get; }
+
+            [RequiresUnreferencedCode(HttpContentJsonExtensions.SerializationUnreferencedCodeMessage)]
+            [RequiresDynamicCode(HttpContentJsonExtensions.SerializationDynamicCodeMessage)]
+            public JsonContentObjectSerializer(
+                object? inputValue,
+                Type inputType,
+                JsonSerializerOptions? options)
+            {
+                ThrowHelper.ThrowIfNull(inputType);
+
+                if (inputValue != null && !inputType.IsAssignableFrom(inputValue.GetType()))
+                {
+                    throw new ArgumentException(SR.Format(SR.SerializeWrongType, inputType, inputValue.GetType()));
+                }
+
+                Value = inputValue;
+                ObjectType = inputType;
+                _jsonSerializerOptions = options ?? JsonHelpers.s_defaultSerializerOptions;
+            }
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
```

Comment:
Once you remove the `JsonSerializerOptions`, these suppressions are no longer necessary since you'll be calling into the `JsonTypeInfo` overloads.

Reply:
`JsonContentObjectSerializer ` non-JsonTypeInfo pattern, which requires unreferenced code. This logic was simply moved from the previous implementation of `JsonContent`. It doesn't have access to a `JsonTypeInfo`.

Reply:
Per https://github.com/dotnet/runtime/pull/89614#discussion_r1277502557 you can expression `JsonSerializerOptions` in terms of `JsonTypeInfo`. This should eliminate most of the code duplication.

Reply:
I'm struggling a bit to follow this, can you explain a bit more? If the consumer calls the non-AOT compatible `JsonContent.Create(object? inputValue, Type inputType, MediaTypeHeaderValue? mediaType = null, JsonSerializerOptions? options = null)` endpoint and we use the default reflection-based `JsonSerializerOptions` I'm getting a `NotSupportedException` saying `JsonTypeInfo metadata for type 'System.Net.Http.Json.Functional.Tests.Person' was not provided by TypeInfoResolver of type '<null>'. If using source generation, ensure that all root types passed to the serializer have been annotated with 'JsonSerializableAttribute', along with any types that might be serialized polymorphically.` if I call `GetTypeInfo`.

Reply:
For the methods that accept a nullable `JsonSerializerOptions?` parameter you need to initialize/normalize the parameter as follows:
```C#
options ??= JsonSerializerOptions.Default;
options.TypeInfoResolver ??= JsonSerializerOptions.Default.TypeInfoResolver;
options.MakeReadOnly();
```
That preamble should ensure you'll get the exepected behavior when calling `GetTypeInfo` later.

Reply:
Thanks, that did the trick. I've updated to always use `JsonTypeInfo` now.